### ISSUE TITLE
access_by_lua_file does not work

### ISSUE THREAD

**Issue by wangpre** [openresty/lua-nginx-module#2176]
When using access_by_lua_file and proxy_pass together, it does not work.
```
location /v1/ {
        access_by_lua_file /data/lua/access.lua;

        proxy_pass       https://api.xxx.com;
        proxy_set_header Host api.xxx.com;
        proxy_set_header X-Real-IP $remote_addr;
    }
```
access.lua
```

local redis = require "resty.redis"

local red = redis:new()
red:set_timeout(1000)
local ok, err = red:connect("127.0.0.1", 6379)
if not ok then
    ngx.log(ngx.ERR, "failed to connect to Redis: ", err)
    return
end

local client_ip = ngx.var.remote_addr

local new_count, err = red:zincrby("ip_counts", 1, client_ip)
if not new_count then
    ngx.log(ngx.ERR, "failed to increase count for ", client_ip, ": ", err)
end

red:close()
```

**Comment by oowl** [comment]
Can you give us the small reproduce code to debug this? It seems not the correct behavior.

**Comment by tYYGH** [comment]
Hello, I confirm the bug. I compiled openresty 1.21.4.2 today, in order to use `init_by_lua_file` (called in the `http` block) and `access_by_lua_file` (I tried with it being called in the `http` or `server` block).
It works well as long as no `location` has a `proxy_pass`; as soon as one such `location` appears, the `access_by_lua_file` instruction gets ignored.

**Comment by lynch1981** [comment]
I tried to reproduce this issue using openresty 1.21.4.2, but it seems to work fine. 
so can you give me the small reproduce code to debug this?

**Comment by tYYGH** [comment]
Thank you for looking, lynch1981. I will try to shrink my conf to a minimal setup that reproduces the issue, and I’ll report it.

**Comment by lynch1981** [comment]
thanks , please note that access_by_lua_file does not run in subrequests

**Comment by tYYGH** [comment]
The issue is apparently more subtle than I thought…

For the record, on my server, Haproxy listens on 80 and 443, and terminates TLS. Then , Nginx receives traffic from 80 and 443 in unencrypted form through 2 Unix sockets in proxy-protocol.

I wrote this “minimal” `nginx.conf` file, and… `curl -ki https://localhost/` did return 418 :-/

```
worker_processes auto;
error_log  /var/log/nginx/error.log  debug;

events {
  worker_connections  1024;
}

http {
  include     mime.types;
  default_type  application/octet-stream;
  sendfile    on;
  keepalive_timeout  65;

  lua_shared_dict  cache 10m;
  init_by_lua_file /etc/nginx/ssso/do_init.lua;

  client_body_temp_path /var/tmp/client_body_temp;
  proxy_temp_path /var/tmp/proxy_temp;
  fastcgi_temp_path /var/tmp/fastcgi_temp;
  uwsgi_temp_path /var/tmp/uwsgi_temp;
  client_max_body_size 10000M;
  gzip on;
  gzip_comp_level 6;
  gzip_proxied any;
  gzip_types application/atom+xml application/javascript application/json application/ld+json application/manifest+json application/rss+xml application/vnd.geo+json application/vnd.ms-fontobject application/x-
font-ttf application/x-web-app-manifest+json application/xhtml+xml application/xml font/opentype image/bmp image/svg+xml image/x-icon text/cache-manifest text/css text/javascript text/plain text/vcard text/vnd
.rim.location.xloc text/vtt text/x-component text/x-cross-domain-policy text/xml;
  index index.php index.html;
  log_format proxy_log '$proxy_protocol_addr - $remote_user [$time_local] "$request" $status $body_bytes_sent "$http_referer" "$http_user_agent"';
  reset_timedout_connection on;
  server_tokens off;
  root /srv/http;

  server {
    listen unix:/run/nginx/http.pp proxy_protocol;
    access_log /var/log/nginx/http_access.log proxy_log;
    set_real_ip_from unix:;
    real_ip_header proxy_protocol;
    set $proxy_scheme "http";
    set $proxy_https "off";
    set $proxy_port "80";

    access_by_lua_block {
      ngx.exit(202)
    }

    location / {
    }
  }

  server {
    listen unix:/run/nginx/https.pp proxy_protocol;
    access_log /var/log/nginx/https_access.log proxy_log;
    set_real_ip_from unix:;
    real_ip_header proxy_protocol;
    set $proxy_scheme "https";
    set $proxy_https "on";
    set $proxy_port "443";

    access_by_lua_block {
      ngx.exit(418)
    }

    location / {
      proxy_pass http://unix:/run/nginx/http.pp:/;
      proxy_set_header Host $host;
      proxy_set_header X-Real-IP $remote_addr;
    }
  }
}
```

I’ll try and pinpoint the exact issue…

**Comment by tYYGH** [comment]
I’ve got to a point where I have a “OK” `nginx.conf` that works, and a “KO” `nginx.conf` that does not, and these two configurations, as reported by `/opt/openresty/nginx/sbin/nginx -T`, are almost identical:
* Include depths differ. Apart from no-brainer (and identical) includes such as mime conf, and PHP conf…, here are the differences:
  KO has a minimal `nginx.conf` that includes the lua-init part and the `server` parts, which themselves include the lua-access part and the `location` parts.  
  OK has a fuller `nginx.conf` that contains the lua-init part,  the `server` parts, and the lua-access part, and only includes the `location` parts.  
  However, the end result is exactly the same, when resolving all the includes, apart from these few differences (and comments and empty lines…):
* KO starts by line  
  `include /etc/nginx/main.inc.d/*.inc;`  
  that does nothing, because `main.inc.d` is empty.
* At the place where above example says  
  `access_by_lua_block { \n ngx.exit(418) \n }`  
  OK now includes the exact same `ngx.exit` through line  
  `access_by_lua_file /etc/nginx/do_418.lua;`  
  KO includes instead the intended access-file with line  
  `access_by_lua_file /etc/nginx/ssso/do_access.lua;`

With OK, when I run `curl -ki https://localhost/index.html`, I get HTTP 418.
With KO, when I run `curl -ki https://localhost/index.html`, I get the HTML contents… and I can even write garbage at the start of the `do_access.lua` file: it is totally ignored.

So, I’m still trying to understand…

**Comment by tYYGH** [comment]
Confirmation, and yet **rectification**, of above.
In my real configuration, `/index.html` is not intended to be hindered by `do_access.lua` thus it is normal that I get the HTML output. My bad.

However, I took again the output of `/opt/openresty/nginx/sbin/nginx -T` for OK and KO configuration, then:
* removed `include …/main.d/*…` since the folder is empty;
* moved the `# configuration file …/conf.d/…` contents as a substitute for `include …/conf.d/*…` lines;
* etc. in order to have _only_ `location`-related `…/inc.d/*.inc` files _included_, in both configurations.

Result:

```bash
# diff -wBbEu0 <(grep -vE '^[[:blank:]]*(#.*)?$' nginx.ko.conf) <(grep -vE '^[[:blank:]]*(#.*)?$' nginx.ok.conf)          
--- /dev/fd/63  2023-10-25 17:56:58.306613259 +0100
+++ /dev/fd/62  2023-10-25 17:56:58.306613259 +0100
@@ -50 +50 @@
-access_by_lua_file        /etc/nginx/ssso/do_access.lua;
+    access_by_lua_file /etc/nginx/do_418.lua;
```

Behaviour:

```bash
# cp nginx.ko.conf nginx.conf

# systemctl restart openresty.service

# curl -ki https://localhost/torrent/web
HTTP/1.1 307 Temporary Redirect
Server: openresty
Date: Wed, 25 Oct 2023 17:03:16 GMT
Content-Type: text/html
Content-Length: 168
Connection: keep-alive
Location: https://mydev.uk/start/login?back=%2Ftorrent%2Fweb&cause=401
Cache-Control: no-cache
…

# cp nginx.ok.conf nginx.conf

# systemctl restart openresty.service

# curl -ki https://localhost/torrent/web
HTTP/1.1 418 
Server: openresty
Date: Wed, 25 Oct 2023 16:50:32 GMT
Content-Length: 0
Connection: keep-alive
```

In short, **both behave as expected**. So, unless I just was initially mistaken (in which case I’ll feel stupid), the only reason I can see for having a good result now, and not before, would be a difference in `include` levels.

That’s worth some more testing…

**Comment by tYYGH** [comment]
Well… I can feel stupid now :laughing: because OpenResty is working perfectly. Sorry for the noise.

I was thrown away by results which I did not understand, and which seemed to fit this “bug” of which I found several occurrences in search results… (probably people like me who did not completely understand Nginx inner working).

The important thing to understand, and which explained everything, is that locations that immediately return an HTTP code are run _before access-by-lua_. This encompasses `return xxx;` statements, as well as `rewrite regex new-url;` statements.  
I don’t know if “_`access_by_lua_file` does not run in subrequests_”, which I’ve read again and again, covers this behaviour, but if it does, it’s certainly not clear for new-comers to Nginx.

**Comment by lynch1981** [comment]
README.markdown of ngx_lua are always worth reading:)
